File: src/637-average-level-binary-tree/java/AverageLevelBinaryTree.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;

public class AverageLevelBinaryTree {
    public List<Double> averageOfLevels(TreeNode root) {
        Queue<TreeNode> treeQueue = new LinkedList<>(List.of(root));
        List<Double> ans = new ArrayList<>();
        while (treeQueue.size() > 0) {
            double qlen = treeQueue.size(), rowSum = 0;
            for (int i = 0; i < qlen; i++) {
                TreeNode currNode = treeQueue.poll();
                rowSum += currNode.val;
                if (currNode.left != null)
                    treeQueue.offer(currNode.left);
                if (currNode.right != null)
                    treeQueue.offer(currNode.right);
            }
            ans.add(rowSum / qlen);
        }
        return ans;
    }

    class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
}

File: src/637-average-level-binary-tree/kotlin/AverageLevelBinaryTree.kt
class AverageLevelBinaryTree {
    public List<Double> averageOfLevels(TreeNode root) {
        if(root == null)
            return doubleArrayOf()
        val treeQueue : Queue<TreeNode> = LinkedList()
        treeQueue.offer(root)
        val ans = mutableListOf<Double>()
        while(treeQueue.isNotEmpty()) {
            val size = treeQueue.size
            var sum = 0.0
            for(i in 0 until size) {
                val node = treeQueue.poll()
                sum += node.`val`.toDouble()
                node.left?.let{ treeQueue.offer(it) }
                node.right?.let{ treeQueue.offer(it) }
            }
            ans.add(sum.toDouble() / size.toDouble())
        }
        return ans.toDoubleArray()
    }
    }
    
    class TreeNode(var `val`: Int) {
        var left: TreeNode? = null
        var right: TreeNode? = null
    }
}